eventtracking/django/models.py

```diff
@@ -0,0 +1,171 @@
+    invalid_regex_expressions = []
+    valid_regex_expressions = []
+    for exp in expressions_list:
+        try:
+            re.compile(exp)
+            valid_regex_expressions.append(exp)
+        except re.error:
+            invalid_regex_expressions.append(exp)
+    return valid_regex_expressions, invalid_regex_expressions
+
+
+class RegExFilter(TimeStampedModel):
+    """
+    This filter uses regular expressions to filter the events
+    """
+
+    BLACKLIST = 'blacklist'
+    WHITELIST = 'whitelist'
+    FILTER_TYPES = (
+        (BLACKLIST, 'Blacklist'),
+        (WHITELIST, 'Whitelist'),
+    )
+
+    backend_name = models.CharField(
+        max_length=50,
+        verbose_name='Backend name',
+        null=False,
+        blank=False,
+        help_text=(
+            'Name of the tracking backend on which this filter should be applied.'
+            '<br/>'
+            'Please note that this field is <b>case sensitive.</b>'
+        )
+    )
+
+    is_enabled = models.BooleanField(
+        default=True,
+        verbose_name='Is Enabled'
+    )
+
+    filter_type = models.CharField(
+        max_length=9,
+        choices=FILTER_TYPES,
+        verbose_name='Filter type',
+        help_text=(
+            'Whitelist: Only events matching the regular expressions in the list '
+            'will be allowed to passed through.'
+            '<br/>'
+            'Blacklist: Events matching any regular expression in the list will be '
+            'blocked.')
+    )
+
+    regular_expressions = models.TextField(
+        max_length=500,
+        verbose_name='List of regular expressions',
+        help_text=('This should be a list of regular expressions, seperated '
+                   'by a newline, for the events to be filtered.'),
+        validators=(validate_regex_list, )
+    )
+
+    class Meta:
+        ordering = ('backend_name', 'is_enabled', '-modified')
+
+    def __str__(self):
+        return '{backend} - {type} - {is_enabled}'.format(
+            backend=self.backend_name, type=self.filter_type, is_enabled=self.is_enabled
+        )
+
+    @property
+    def compiled_expressions(self):
+        """
+        Return a list of compiled regular expressions
+        """
+        valid_expressions, invalid_expressions = _clean_expressions(self.regular_expressions)
+        if invalid_expressions:
+            logger.error(
+                'The following invalid expressions were found in the filter: %s',
+                ', '.join(invalid_expressions)
+            )
+        return [re.compile(exp) for exp in valid_expressions]
+
+    @classmethod
+    def get_latest_enabled_filter(cls, backend_name=None):
+        """
+        Return the last modified filter.
+
+        If backend_name is provided, search through filters matching the backend_name
+        otherwise search among all available filters.
+
+        Return None if there is no filter exists that matches the criteria.
+        """
+        if backend_name:
```

Comment:
can we change it to 
```
queryset = cls.objects.filter(is_enabled=True)
if backend_name:
    queryset = cls.objects.filter(backend_name=backend_name)
return queryset.latest('modified').first()
```

Reply:
I'll modify the `backend_name` check.
but we need the if else for `latest` query since for empty query set, `latest` returns the actual object, not a queryset, and raises an exception for an empty queryset
```
>>> RegExFilter.objects.filter(is_enabled=True, backend_name='asd').latest('modified')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/django/db/models/query.py", line 649, in latest
    return self.reverse()._earliest(*fields, field_name=field_name)
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/django/db/models/query.py", line 643, in _earliest
    return obj.get()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/django/db/models/query.py", line 408, in get
    self.model._meta.object_name
eventtracking.django.models.RegExFilter.DoesNotExist: RegExFilter matching query does not exist.
>>> RegExFilter.objects.filter(is_enabled=True, backend_name='caliper').latest('modified')
<RegExFilter: caliper - blacklist - True>
>>>
```